Readme.md

# Pay with Amazon

Simple subscription billing with Amazon!

[Demo][demo]

[Download][download]

## Introduction

This JavaScript library makes it easy to add Pay with Amazon to your payments
flow. It wraps Amazon's Login and Pay with Amazon JavaScript libraries to
simplify its integration. See [Amazon's introduction to this service][amazon-docs]
for more detail.

Hundreds of millions of Amazon customers can subscribe to your subscription
plans and make purchases using their Amazon account.

This library is platform agnostic, and may be used in any checkout flow to
make Amazon Payments easier to implement. To learn how use this library with
your Recurly site, please see [Recurly's documentation][recurly-docs] to get started.

## Usage

Simply include `pay-with-amazon.min.js` in your document head.

**Important!** This **must** be included in the `<head>` of your document.
[[1](#1)]

```html
<head>
  ...
  <script src="pay-with-amazon.min.js"></script>
  ...
</head>
```

Then, also in the document `<head>`, invoke `PayWithAmazon`. Each widget
property accepts a String corresponding to the `id` of an element on your
page. These elements are where the corresponding widget will be injected.

For example:

```html
<head>
  ...
  <script>
    var payWithAmazon = new PayWithAmazon({
      sellerId: 'ABC',
      clientId: 'XYZ',
      button: 'pay-with-amazon',
      addressBook: 'address-book',
      wallet: 'wallet',
      consent: 'consent'
    });
  </script>
  ...
</head>
```

Each widget option also accepts an object value to pass more options.

```js
var payWithAmazon = new PayWithAmazon({
  sellerId: 'ABC',
  clientId: 'XYZ',
  button: { id: 'pay-with-amazon', [type], [color] },
  addressBook: { id: 'address-book', [width], [height] },
  wallet: { id: 'wallet', [width], [height] },
  consent: { id: 'consent', [width], [height] }
});
```

To add EU support, simply include the `region` parameter:

```js
var payWithAmazon = new PayWithAmazon({
  sellerId: 'ABC',
  clientId: 'XYZ',
  button: { id: 'pay-with-amazon', [type], [color] },
  addressBook: { id: 'address-book', [width], [height] },
  wallet: { id: 'wallet', [width], [height] },
  consent: { id: 'consent', [width], [height] },
  region: 'eu'
});
```

For this example, the `<body>` would need to contain the following elements:

```html
<body>
  ...
  <div id="pay-with-amazon"></div>
  <div id="address-book"></div>
  <div id="wallet"></div>
  <div id="consent"></div>
  ...
</body>
```

### Styling

If widget dimensions are not passed in directly, PayWithAmazon will rely
upon the dimensions of their container elements.

PayWithAmazon will add a class to each widget container, `'open'` by default,
but customizable by the `openedClass` option. This allows more robust styling
than simply specifying dimensions. For an example of how this enables the
widgets to be sized responsively, see the [responsive demo][responsive-demo]

## API

### PayWithAmazon(opts)

Initializes a new PayWthAmazon object.

Param | Type | Description
----- | ---- | -----------
sellerId | String | Amazon Seller ID
clientId | String | Amazon Client ID
[production] | Boolean | Whether to use production widgets (defaults to false)
button | Object/String | [button options](#button-options). If String, sets button.id
wallet | Object/String | [wallet options](#wallet-options). If String, sets wallet.id
[addressBook] | Object/String | [addressBook options](#addressbook-options). If excluded, customers will interact solely with the wallet and consent widgets. If String, sets addressBook.id
[consent] | Object/String | [conset options](#consent-options). If excluded, customers will not need to give a consent. If String, sets consent.id
[openedClass] | String | Class name to add to containers once their widgets have been opened. Defaults to 'open'

#### button options

Param | Type | Description
----- | ---- | -----------
id | String | DOM node id in which the Amazon login button will be placed
[type] | String | 'large' (default), 'small' [See button guide][button-guide]
[color] | String | 'Gold' (default), 'LightGray', 'DarkGray' [See button guide][button-guide]
[kind] | String | 'pay' (default), 'login' [See button guide][button-guide]

#### wallet options

Param | Type | Description
----- | ---- | -----------
id | String | DOM node id in which the Amazon wallet widget will be placed
[width] | Number | Width of the widget in pixels
[height] | Number | Height of the widget in pixels

#### addressBook options

Param | Type | Description
----- | ---- | -----------
id | String | DOM node id in which the Amazon address book widget will be placed
[width] | Number | Width of the widget in pixels
[height] | Number | Height of the widget in pixels

#### consent options

Param | Type | Description
----- | ---- | -----------
id | String | DOM node id in which the Amazon consent widget will be placed
[width] | Number | Width of the widget in pixels
[height] | Number | Height of the widget in pixels

### #on(event, listener)

Attaches a listener to the specified event. See below for possible
event names.

### #off(event, listener)

Detaches a listener from the specified event.

### Events

A PayWithAmazon instance may emit events during setup and customer
interaction with the Amazon widgets.

Listeners can be attached with the `on` method, and detached with
the `off` method.

#### change

Whenever the customer makes a change to their billing agreement.

**Emits:** An object describing the customer's status.

```js
{
  id: 'abc-xyz', // Amazon billing agreement id.
  consent: true // Whether billing consent has been given
}
```

If the consent is not required:

```js
{
  id: 'abc-xyz' // Amazon billing agreement id.
}
```

And in case of an error:

```js
{
  consent: false, // May be true, false or not present if the consent is not required
  error: 'Billing consent not given.' // A brief description of the error
}
```

#### login

Whenever the customer logs in. This may be used to hide the login button, or
perform other page transitions.

**Emits:** Nothing.

#### ready.addressBook

The address book widget has finished loading and is now displayed on the page.

**Emits:** Nothing.

#### ready.wallet

The wallet widget has finished loading and is now displayed on the page.

**Emits:** Nothing.

#### ready.consent

The consent widget has finished loading and is now displayed on the page.

**Emits:** Nothing.

#### error

Whenever an error occurs setting up the widgets or while the customer is
interacting with the Amazon widgets.

**Emits:** and object describing the error.

```js
{
  code: 'InvalidSellerId', // A string representing an error code
  message: 'The seller identifier that you have provided is invalid. Specify a valid SellerId.' // A brief description of the error
}
```

[See Amazon's documentation][error-codes] for a list of all possible error codes.

## Use Cases

### Case 1: Standard Subscriptions

This is a basic example of the standard settings for the Amazon Payment
widgets. This would display the Amazon address, payment and consent widgets.

```js
var payWithAmazon = new PayWithAmazon({
  sellerId: 'ABC',
  clientId: 'XYZ',
  button: { id: 'pay-with-amazon', type: 'large', color: 'DarkGray' },
  addressBook: { id: 'address-book', width: 400, height: 260 },
  wallet: { id: 'wallet', width: 400, height: 260 },
  consent: { id: 'consent', width: 400, height: 140 }
});
```

### Case 2: No Address Widget

You may want to streamline your checkout experience if you are selling only
digital goods and do not require an account address. In this case, you can
remove the Amazon address widget.

```js
var payWithAmazon = new PayWithAmazon({
  sellerId: 'ABC',
  clientId: 'XYZ',
  button: { id: 'pay-with-amazon', type: 'large', color: 'DarkGray' },
  wallet: { id: 'wallet', width: 400, height: 260 },
  consent: { id: 'consent', width: 400, height: 140 }
});
```

### Case 3: No Consent Widget

You may not want the customer consent in your checkout experience. In this case, you can remove the Amazon Consent widget.

```js
var payWithAmazon = new PayWithAmazon({
  sellerId: 'ABC',
  clientId: 'XYZ',
  button: { id: 'pay-with-amazon', type: 'large', color: 'DarkGray' },
  addressBook: { id: 'address-book', width: 400, height: 260 },
  wallet: { id: 'wallet', width: 400, height: 260 },
});
```

## Contributing

[See Contributing file][contributing]

## Compatibility

IE11+, Firefox, Chrome, Safari

## Notes

##### 1
Amazon's widgets use methods that will otherwise fail if they are not
invoked from the document `<head>`.

## License

The MIT License (MIT)

Copyright (c) 2020 Recurly, inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

[demo]: https://recurly.github.io/pay-with-amazon/
[responsive-demo]: https://recurly.github.io/pay-with-amazon/responsive.html
[download]: https://github.com/recurly/pay-with-amazon/releases
[error-codes]: https://payments.amazon.com/developer/documentation/lpwa/201954960
[button-guide]: https://payments.amazon.com/developer/documentation/lpwa/201953980
[amazon-docs]: https://payments.amazon.com/developer/documentation/lpwa/201985870
[recurly-docs]: https://docs.recurly.com/payment-gateways/amazon-payments
[contributing]: CONTRIBUTING.md


CHANGELOG.md
# Changelog

## [v2.1.1](https://github.com/recurly/pay-with-amazon/tree/v2.1.1) (2022-06-08)

[Full Changelog](https://github.com/recurly/pay-with-amazon/compare/v2.1.0...v2.1.1)


**Merged Pull Requests**

- Consent widget is optional [#39](https://github.com/recurly/pay-with-amazon/pull/39) ([cyberxander90](https://github.com/cyberxander90))
- Revert "add security to github CODEOWNER file" [#33](https://github.com/recurly/pay-with-amazon/pull/33) ([eternal44](https://github.com/eternal44))
- add security to github CODEOWNER file [#32](https://github.com/recurly/pay-with-amazon/pull/32) ([eternal44](https://github.com/eternal44))

**Closed Issues**

- Can't render small button [#5](https://github.com/recurly/pay-with-amazon/issues/5)


## [v2.1.0](https://github.com/recurly/pay-with-amazon/tree/v2.1.0) (2021-04-21)

[Full Changelog](https://github.com/recurly/pay-with-amazon/compare/v2.0.1...v2.1.0)


**Merged Pull Requests**

- Adding github action to publish demo pages. [#29](https://github.com/recurly/pay-with-amazon/pull/29) ([douglasmiller](https://github.com/douglasmiller))
- Accept additional authorization scopes [#28](https://github.com/recurly/pay-with-amazon/pull/28) ([jguidry-recurly](https://github.com/jguidry-recurly))

**Closed Issues**

- Can't render small button [#5](https://github.com/recurly/pay-with-amazon/issues/5)


## [v2.0.1](https://github.com/recurly/pay-with-amazon/tree/v2.0.1) (2020-07-13)

[Full Changelog](https://github.com/recurly/pay-with-amazon/compare/v2.0.0...v2.0.1)


**Merged Pull Requests**

- Fixes asset path determination [#21](https://github.com/recurly/pay-with-amazon/pull/21) ([chrissrogers](https://github.com/chrissrogers))

**Closed Issues**

- _this.getAssetPath is not a function [#20](https://github.com/recurly/pay-with-amazon/issues/20)


## [v2.0.0](https://github.com/recurly/pay-with-amazon/tree/v2.0.0) (2020-03-20)

[Full Changelog](https://github.com/recurly/pay-with-amazon/compare/v1.0.4...v2.0.0)


**Merged Pull Requests**

- V2 [#17](https://github.com/recurly/pay-with-amazon/pull/17) ([chrissrogers](https://github.com/chrissrogers))
- Add support for Amazon Pay UK [#14](https://github.com/recurly/pay-with-amazon/pull/14) ([jpgnotgif](https://github.com/jpgnotgif))
- Add support for Amazon Pay EU [#12](https://github.com/recurly/pay-with-amazon/pull/12) ([jpgnotgif](https://github.com/jpgnotgif))

**Closed Issues**

- sync loading warning message [#16](https://github.com/recurly/pay-with-amazon/issues/16)
- Add support for responsive widgets via per-widget flag [#2](https://github.com/recurly/pay-with-amazon/issues/2)


## [v1.0.4](https://github.com/recurly/pay-with-amazon/tree/v1.0.4) (2017-02-13)

[Full Changelog](https://github.com/recurly/pay-with-amazon/compare/v1.0.3...v1.0.4)


**Merged Pull Requests**

- Fix OffAmazonPayments check; emit login information [#7](https://github.com/recurly/pay-with-amazon/pull/7) ([keelerm84](https://github.com/keelerm84))

**Closed Issues**

- Using with AngularJS [#6](https://github.com/recurly/pay-with-amazon/issues/6)
- Non-recurring? [#4](https://github.com/recurly/pay-with-amazon/issues/4)


## [v1.0.3](https://github.com/recurly/pay-with-amazon/tree/v1.0.3) (2014-12-25)

[Full Changelog](https://github.com/recurly/pay-with-amazon/compare/v1.0.2...v1.0.3)





## [v1.0.2](https://github.com/recurly/pay-with-amazon/tree/v1.0.2) (2014-10-09)

[Full Changelog](https://github.com/recurly/pay-with-amazon/compare/v1.0.1...v1.0.2)





## [v1.0.1](https://github.com/recurly/pay-with-amazon/tree/v1.0.1) (2014-09-17)

[Full Changelog](https://github.com/recurly/pay-with-amazon/compare/v1.0.0...v1.0.1)





## [v1.0.0](https://github.com/recurly/pay-with-amazon/tree/v1.0.0) (2014-09-03)






webpack.config.js
const TerserPlugin = require('terser-webpack-plugin');
const path = require('path');
const minify = ~process.argv.indexOf('-p');
const manifest = require('./package.json');

module.exports = {
  node: {
    global: false,
  },
  entry: './index',
  mode: 'development',
  output: {
    path: path.join(__dirname, 'build'),
    publicPath: '/build/',
    filename: 'pay-with-amazon' + (minify ? '.min.js' : '.js'),
    library: 'PayWithAmazon',
    libraryExport: 'default'
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: [
          {
            loader: 'babel-loader',
            options: {
              cacheDirectory: true,
              plugins: [
                '@babel/plugin-proposal-class-properties',
              ],
              presets: [
                ['@babel/preset-env', {
                  targets: {
                    browsers: manifest.browserslist
                  }
                }]
              ]
            }
          }
        ]
      }
    ]
  },
  resolve: {
    extensions: ['.js'],
    modules: [
      path.join(__dirname),
      'node_modules'
    ]
  },
  optimization: {
    minimizer: [
      new TerserPlugin({
        terserOptions: {
          output: {
            comments: false,
          },
        },
        extractComments: false
      })
    ]
  },
  devServer: {
    disableHostCheck: true
  }
};


pay-with-amazon.responsive.js
(function outer(modules, cache, entries){

  /**
   * Global
   */

  var global = (function(){ return this; })();

  /**
   * Require `name`.
   *
   * @param {String} name
   * @param {Boolean} jumped
   * @api public
   */

  function require(name, jumped){
    if (cache[name]) return cache[name].exports;
    if (modules[name]) return call(name, require);
    throw new Error('cannot find module "' + name + '"');
  }

  /**
   * Call module `id` and cache it.
   *
   * @param {Number} id
   * @param {Function} require
   * @return {Function}
   * @api private
   */

  function call(id, require){
    var m = cache[id] = { exports: {} };
    var mod = modules[id];
    var name = mod[2];
    var fn = mod[0];

    fn.call(m.exports, function(req){
      var dep = modules[id][1][req];
      return require(dep ? dep : req);
    }, m, m.exports, outer, modules, cache, entries);

    // expose as `name`.
    if (name) cache[name] = cache[id];

    return cache[id].exports;
  }

  /**
   * Require all entries exposing them on global if needed.
   */

  for (var id in entries) {
    if (entries[id]) {
      global[entries[id]] = require(id);
    } else {
      require(id);
    }
  }

  /**
   * Duo flag.
   */

  require.duo = true;

  /**
   * Expose cache.
   */

  require.cache = cache;

  /**
   * Expose modules
   */

  require.modules = modules;

  /**
   * Return newest require.
   */

   return require;
})({
1: [function(require, module, exports) {
'use strict';

/**
 * Dependencies
 */

var Emitter = require('component/emitter');
var bind = require('component/bind');

/**
 * Exports
 */

module.exports = PayWithAmazon;

/**
 * Off Amazon Payments wrapper
 *
 * order of operations
 *
 *   button -> address book -> wallet -> consent
 *
 * example
 *
 *   new PayWithAmazon({
 *     sellerId: 'abc',
 *     clientId: 'xyz',
 *     button: { id: 'pay-with-amazon' },
 *     addressBook: { id: 'address-book' [, width: 400 [, height: 260]]},
 *     wallet: { id: 'wallet' [, width: 400 [, height: 260]]},
 *     consent: { id: 'consent' [, width: 400 [, height: 140]]}
 *   });
 *
 * @param {Object} opts
 * @param {String} opts.sellerId
 * @param {String} opts.clientId
 * @param {Object|String} opts.button if string, sets opts.button.id
 * @param {String} opts.button.id
 * @param {String} [opts.button.type] 'large' (default), 'small'
 * @param {String} [opts.button.color] 'Gold' (default), 'LightGray', 'DarkGray'
 * @param {Object|String} opts.wallet
 * @param {String} opts.wallet.id
 * @param {Number} [opts.wallet.width]
 * @param {Number} [opts.wallet.height]
 * @param {Object|String} [opts.addressBook]
 * @param {String} opts.addressBook.id
 * @param {Number} [opts.addressBook.width]
 * @param {Number} [opts.addressBook.height]
 * @param {Object|String} opts.consent
 * @param {String} opts.consent.id
 * @param {Number} [opts.consent.width]
 * @param {Number} [opts.consent.height]
 * @param {String} [opts.openedClass]
 */

function PayWithAmazon (opts) {
  if (!(this instanceof PayWithAmazon)) return new PayWithAmazon(opts);

  this.configure(opts);

  this.billingAgreementId = null;
  this.consent = undefined;
  this.widgets = {};
  this._status = this.status();

  this.setBillingAgreementId = bind(this, this.setBillingAgreementId);
  this.initWallet = bind(this, this.initWallet);
  this.initConsent = bind(this, this.initConsent);
  this.setConsent = bind(this, this.setConsent);
  this.error = bind(this, this.error);
  this.init = bind(this, this.init);

  if ('window.OffAmazonPayments' in window) {
    this.init();
  } else {
    document.write('<script src="'
      + 'https://static-na.payments-amazon.com/OffAmazonPayments'
      + '/us/sandbox/js/Widgets.js?sellerId='
      + this.config.sellerId
      + '"></script>');

    window.onAmazonLoginReady = this.init;
  }

  return this;
}

/**
 * PayWithAmazon inherits Emitter
 */

Emitter(PayWithAmazon.prototype);

/**
 * Version
 */

PayWithAmazon.prototype.version = '1.0.1';

/**
 * Configures the instance based on passed `opts`
 *
 * Throws errors if the opts are insufficient to configure the instance,
 * and sets defaults for those not specified.
 *
 * See initializer for description of `opts`
 */

PayWithAmazon.prototype.configure = function (opts) {
  if (typeof opts !== 'object') throw new Error ('opts must be provided as an object.');

  if (!opts.sellerId) throw new Error('opts.sellerId is required.');
  if (!opts.clientId) throw new Error('opts.clientId is required.');
  if (!opts.button) throw new Error('opts.button is required.');
  if (!opts.wallet) throw new Error('opts.wallet is required.');
  if (!opts.consent) throw new Error('opts.consent is required.');

  if (typeof opts.button === 'string') opts.button = { id: opts.button };
  if (typeof opts.wallet === 'string') opts.wallet = { id: opts.wallet };
  if (typeof opts.consent === 'string') opts.consent = { id: opts.consent };
  if (typeof opts.addressBook === 'string') opts.addressBook = { id: opts.addressBook };

  opts.button.type = opts.button.type === 'small' ? 'Pay' : 'PwA';
  opts.button.color = opts.button.color || 'Gold';

  if (opts.wallet.width || opts.wallet.height) {
    opts.wallet.dimensions = {
      width: dimension(opts.wallet.width || 400),
      height: dimension(opts.wallet.height || 260)
    };
  }

  if (opts.consent.width || opts.consent.height) {
    opts.consent.dimensions = {
      width: dimension(opts.consent.width || 400),
      height: dimension(opts.consent.height || 140)
    };
  }

  if (opts.addressBook && (opts.addressBook.width || opts.addressBook.height)) {
    opts.addressBook.dimensions = {
      width: dimension(opts.addressBook.width || 400),
      height: dimension(opts.addressBook.height || 260)
    };
  }

  opts.openedClass = opts.openedClass || 'open';

  this.config = opts;
};

/**
 * Initialized the Amazon plugin
 *
 * Sets the client ID then waits for the widget classes to load
 * before initializing the login button.
 */

PayWithAmazon.prototype.init = function () {
  window.amazon.Login.setClientId(this.config.clientId);

  var self = this;
  var pollId = setInterval(poll, 200);

  function poll () {
    if (!window.OffAmazonPayments.Button) return;
    clearTimeout(pollId);
    self.initButton();
  }
};

/**
 * Returns an object describing the customer state
 *
 * @return {Object} customer state
 */

PayWithAmazon.prototype.status = function () {
  var id = this.billingAgreementId;
  var consent = this.consent;
  var status = {};

  if (!id) {
    status.error = 'Billing agreement ID has not been set.';
  }

  if (consent !== undefined) {
    status.consent = consent;
  }

  if (consent === undefined) {
    status.error = 'Billing consent not yet given.';
  } else if (!consent) {
    status.error = 'Billing consent not given.';
  }

  if (!status.error) {
    status.id = id;
  }

  return status;
};

/**
 * Emits the 'change' event if the customer status has changed
 */

PayWithAmazon.prototype.check = function () {
  var status = this.status();
  if (JSON.stringify(status) !== JSON.stringify(this._status)) {
    this._status = status;
    this.emit('change', status);
  }
};

/**
 * Initializes the login button
 */

PayWithAmazon.prototype.initButton = function () {
  var self = this;
  var type = this.config.button.type;
  var color = this.config.button.color;

  this.widgets.button = new window.OffAmazonPayments.Button(this.config.button.id, this.config.sellerId, {
    type: type,
    color: color,
    authorization: function () {
      var opts = {
        scope: 'profile payments:widget payments:shipping_address',
        popup: true
      };

      window.amazon.Login.authorize(opts, function (res) {
        if (res.error) return self.error(res.error);
        self.emit('login');
        self.initAddressBook();
      });
    },
    onError: this.error
  });
};

/**
 * Initializes the Address Book widget
 */

PayWithAmazon.prototype.initAddressBook = function () {
  var self = this;

  if (!this.config.addressBook) return this.initWallet();

  var opts = {
    agreementType: 'BillingAgreement',
    sellerId: this.config.sellerId,
    onReady: function (ref) {
      self.emit('ready.addressBook');
      self.setBillingAgreementId(ref);
    },
    onAddressSelect: this.initWallet,
    design: design(this.config.addressBook),
    onError: this.error
  };

  this.widgets.addressBook = new window.OffAmazonPayments.Widgets.AddressBook(opts);
  this.widgets.addressBook.bind(this.config.addressBook.id);
  this.opened(this.config.addressBook.id);
};

/**
 * Initializes the wallet widget
 */

PayWithAmazon.prototype.initWallet = function () {
  var self = this;
  var opts = {
    amazonBillingAgreementId: this.billingAgreementId,
    sellerId: this.config.sellerId,
    design: design(this.config.wallet),
    onReady: function (ref) {
      self.emit('ready.wallet');
      if (!self.billingAgreementId) {
        self.setBillingAgreementId(ref);
      }
    },
    onPaymentSelect: function () {
      self.initConsent();
      self.check();
    },
    onError: this.error
  };

  if (!this.billingAgreementId) {
    opts.agreementType = 'BillingAgreement';
  }

  this.widgets.wallet = new window.OffAmazonPayments.Widgets.Wallet(opts);
  this.widgets.wallet.bind(this.config.wallet.id);
  this.opened(this.config.wallet.id);
};

/**
 * Initializes the consent widget
 */

PayWithAmazon.prototype.initConsent = function () {
  var self = this;
  var opts = {
    amazonBillingAgreementId: this.billingAgreementId,
    sellerId: this.config.sellerId,
    design: design(this.config.consent),
    onReady: function (consentStatus) {
      self.emit('ready.consent');
      self.setConsent(consentStatus);
    },
    onConsent: this.setConsent,
    onError: this.error
  };

  this.widgets.consent = new window.OffAmazonPayments.Widgets.Consent(opts);
  this.widgets.consent.bind(this.config.consent.id);
  this.opened(this.config.consent.id);
};

/**
 * Sets the billingAgreementId based on a widgit init reference
 */

PayWithAmazon.prototype.setBillingAgreementId = function (ref) {
  this.billingAgreementId = ref.getAmazonBillingAgreementId();
  this.check();
};

/**
 * Sets consent state based on an Amazon ConsentStatus
 */

PayWithAmazon.prototype.setConsent = function (consentStatus) {
  if (typeof consentStatus.getConsentStatus === 'undefined') return;
  this.consent = consentStatus.getConsentStatus() === 'true';
  this.check();
};

/**
 * Adds a class to opened widget containers
 */

PayWithAmazon.prototype.opened = function (id) {
  var elem = document.getElementById(id);
  if (elem) elem.className = elem.className + ' ' + this.config.openedClass;
};

/**
 * Handles errors, logging to console and emitting via the 'error' event
 */

PayWithAmazon.prototype.error = function (err) {
  var error = {
    code: err.getErrorCode(),
    message: err.getErrorMessage()
  };

  if (console) {
    if (Error.captureStackTrace) {
      Error.captureStackTrace(error, this.error);
    }
    console.error(error);
  }

  this.emit('error', error);
};

/**
 * Given a number or string, returns a properly-formatted dimension string
 *
 * @param {Number|String} dim
 * @returns {String}
 */

function dimension (dim) {
  return dim + isNaN(parseInt(dim.charAt(dim.length - 1), 10)) ? '' : 'px';
}

/**
 * Given a widget configuration, returns a properly-formatted design spec
 *
 * @param {Object} widget Widget options hash
 */

function design (widget) {
  if (widget.dimensions) {
    return { size: widget };
  } else {
    return { designMode: 'responsive' };
  }
}

}, {"component/emitter":2,"component/bind":3}],
2: [function(require, module, exports) {

/**
 * Expose `Emitter`.
 */

module.exports = Emitter;

/**
 * Initialize a new `Emitter`.
 *
 * @api public
 */

function Emitter(obj) {
  if (obj) return mixin(obj);
};

/**
 * Mixin the emitter properties.
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */

function mixin(obj) {
  for (var key in Emitter.prototype) {
    obj[key] = Emitter.prototype[key];
  }
  return obj;
}

/**
 * Listen on the given `event` with `fn`.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.on =
Emitter.prototype.addEventListener = function(event, fn){
  this._callbacks = this._callbacks || {};
  (this._callbacks[event] = this._callbacks[event] || [])
    .push(fn);
  return this;
};

/**
 * Adds an `event` listener that will be invoked a single
 * time then automatically removed.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.once = function(event, fn){
  var self = this;
  this._callbacks = this._callbacks || {};

  function on() {
    self.off(event, on);
    fn.apply(this, arguments);
  }

  on.fn = fn;
  this.on(event, on);
  return this;
};

/**
 * Remove the given callback for `event` or all
 * registered callbacks.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.off =
Emitter.prototype.removeListener =
Emitter.prototype.removeAllListeners =
Emitter.prototype.removeEventListener = function(event, fn){
  this._callbacks = this._callbacks || {};

  // all
  if (0 == arguments.length) {
    this._callbacks = {};
    return this;
  }

  // specific event
  var callbacks = this._callbacks[event];
  if (!callbacks) return this;

  // remove all handlers
  if (1 == arguments.length) {
    delete this._callbacks[event];
    return this;
  }

  // remove specific handler
  var cb;
  for (var i = 0; i < callbacks.length; i++) {
    cb = callbacks[i];
    if (cb === fn || cb.fn === fn) {
      callbacks.splice(i, 1);
      break;
    }
  }
  return this;
};

/**
 * Emit `event` with the given args.
 *
 * @param {String} event
 * @param {Mixed} ...
 * @return {Emitter}
 */

Emitter.prototype.emit = function(event){
  this._callbacks = this._callbacks || {};
  var args = [].slice.call(arguments, 1)
    , callbacks = this._callbacks[event];

  if (callbacks) {
    callbacks = callbacks.slice(0);
    for (var i = 0, len = callbacks.length; i < len; ++i) {
      callbacks[i].apply(this, args);
    }
  }

  return this;
};

/**
 * Return array of callbacks for `event`.
 *
 * @param {String} event
 * @return {Array}
 * @api public
 */

Emitter.prototype.listeners = function(event){
  this._callbacks = this._callbacks || {};
  return this._callbacks[event] || [];
};

/**
 * Check if this emitter has `event` handlers.
 *
 * @param {String} event
 * @return {Boolean}
 * @api public
 */

Emitter.prototype.hasListeners = function(event){
  return !! this.listeners(event).length;
};

}, {}],
3: [function(require, module, exports) {
/**
 * Slice reference.
 */

var slice = [].slice;

/**
 * Bind `obj` to `fn`.
 *
 * @param {Object} obj
 * @param {Function|String} fn or string
 * @return {Function}
 * @api public
 */

module.exports = function(obj, fn){
  if ('string' == typeof fn) fn = obj[fn];
  if ('function' != typeof fn) throw new Error('bind() requires a function');
  var args = slice.call(arguments, 2);
  return function(){
    return fn.apply(obj, args.concat(slice.call(arguments)));
  }
};

}, {}]}, {}, {"1":"PayWithAmazon"})
